### DISCUSSION TITLE
Periodic wallpaper rotation

### DISCUSSION THREAD

**Discussion by Sprim56** [end-4/dots-hyprland#724] [Ideas]
To be able to choose a set of wallpapers and have them rotate instead of just one.

**Comment by nihalxkumar** [comment]
This guide will help you set up automatic wallpaper rotation with synchronized theme changes for your Hyprland setup using systemd.

## What This Does

- Changes wallpaper every 30 minutes
- Generates a color scheme based on the new wallpaper
- Updates:
  - AGS bar colors
  - Hyprland colors
  - Terminal colors
  - GTK theme (if using Gradience)
  - Other integrated components

## Prerequisites

Your wallpapers should be in your Pictures directory:
```
~/Pictures/Wallpapers/
```

Your AGS configuration should be in:
```
~/.config/ags/
```

## Setup Steps

1. **Create necessary directories**

```bash
mkdir -p ~/.config/systemd/user/
```

2. **Create the systemd service file**

```bash
nano ~/.config/systemd/user/wallpaper-change.service
```

Add this content:

```ini
[Unit]
Description=Change wallpaper and generate colors periodically
After=graphical-session.target

[Service]
Type=oneshot
Environment=DISPLAY=:0
Environment=WAYLAND_DISPLAY=wayland-1
Environment=XDG_RUNTIME_DIR=/run/user/%U
Environment=PATH=/usr/local/bin:/usr/bin:/bin
Environment=XDG_CONFIG_HOME=%h/.config
Environment=XDG_CACHE_HOME=%h/.cache
Environment=XDG_STATE_HOME=%h/.local/state
Environment=HOME=%h
ExecStart=/bin/bash -c '\
  WALL=$(fd . "$(xdg-user-dir PICTURES)/Wallpapers/" -e .png -e .jpg -e .svg | shuf -n1) && \
  %h/.config/ags/scripts/color_generation/switchwall.sh "$WALL" && \
  sleep 2 && \
  %h/.config/ags/scripts/color_generation/switchcolor.sh "$(cat %h/.local/state/ags/user/color.txt)"'

[Install]
WantedBy=graphical-session.target
```

3. **Create the timer file**
```bash
nano ~/.config/systemd/user/wallpaper-change.timer
```

Add this content:
```ini
[Unit]
Description=Change wallpaper every 30 minutes

[Timer]
OnBootSec=1min
OnUnitActiveSec=30min
Unit=wallpaper-change.service

[Install]
WantedBy=timers.target
```

4. **Enable and start the service**
```bash
# Reload systemd
systemctl --user daemon-reload

# Enable the timer to start at boot
systemctl --user enable wallpaper-change.timer

# Start the timer
systemctl --user start wallpaper-change.timer
```

## Verification

1. **Check if the timer is running**
```bash
systemctl --user status wallpaper-change.timer
```

2. **Test the service manually**
```bash
systemctl --user start wallpaper-change.service
```

3. **Check logs if there are issues**
```bash
journalctl --user -u wallpaper-change.service -f
```

## Customization

- To change the interval, modify `OnUnitActiveSec=30min` in the timer file
- To change wallpaper directory, modify the `fd` command path in the service file
- The `sleep 2` ensures proper color generation; adjust if needed

## Troubleshooting

1. **Service not starting:**
   - Check if all paths in the service file match your setup
   - Verify all required scripts exist in your AGS config
   - Check logs using the journalctl command above

2. **Colors not updating:**
   - Ensure AGS is running
   - Check if color generation scripts have proper permissions
   - Verify the state directory exists: `~/.local/state/ags/`

3. **Wallpaper not changing:**
   - Verify wallpaper directory exists and contains images
   - Check if `fd` command is installed
   - Ensure proper file permissions

## Additional Notes

- The service runs in user space, so no root privileges are required
- The timer will start automatically after boot
- Wallpaper changes are randomized from your wallpaper directory

**Reply by NeoNyxx** [reply]
update??

**Reply by nihalxkumar** [reply]
> update??

Sorry I don’t use hyprland or end4's dots anymore

**Comment by nomaterial** [comment]
This works, but it's a bit outdated for current end-4 setups

You don’t need AGS scripts anymore. Quickshell + switchwall already handle:
wallpaper switching
matugen color generation
syncing Hyprland / GTK / terminal

If your goal is simply to rotate wallpapers from a folder, you can do it much simpler.

1. Create a small script:

#!/usr/bin/env bash
WALL=$(find ~/Pictures/Wallpaper -type f \( -iname "*.jpg" -o -iname "*.png" -o -iname "*.webp" \) | shuf -n 1)
~/.config/quickshell/ii/scripts/colors/switchwall.sh --image "$WALL" --mode dark

2. Then use a systemd user timer:

[Unit]
Description=Rotate wallpaper

[Service]
Type=oneshot
ExecStart=%h/.local/bin/rotate-wallpaper

[Timer]
OnBootSec=2min
OnUnitActiveSec=30min

[Install]
WantedBy=timers.target

3. Enable it:

systemctl --user daemon-reload
systemctl --user enable --now rotate-wallpaper.timer

This gives you:
random wallpaper rotation
automatic color sync
no extra AGS scripts

Tested on Hyprland + quickshell (end-4)
